src/main/java/org/elasticsearch/plugins/repository/DefaultPluginDownloader.java

```diff
@@ -0,0 +1,381 @@
+        while (Files.exists(destination)) {
+            destination = root.resolve(String.valueOf(downloadNumber.getAndIncrement()));
+        }
+
+        DownloadThread dlThread = new DownloadThread(source, destination, listeners, userAgent, elasticsearchVersion);
+        try {
+            dlThread.setDaemon(true);
+            dlThread.start();
+            dlThread.join(timeout.millis());
+
+            if (dlThread.isAlive()) {
+                throw new ElasticsearchTimeoutException("The download operation took longer than " + timeout + ", stopping it.");
+            }
+        } catch (InterruptedException e) {
+            return null;
+        } finally {
+            dlThread.closeStreams();
+        }
+
+        if (dlThread.wasSuccessful()) {
+            return destination;
+        }
+        return null;
+    }
+
+    @Override
+    public void addListener(Listener listener) {
+        listeners.add(listener);
+    }
+
+
+    private class DownloadThread extends Thread {
+
+        private final String source;
+        private final Path destination;
+        private Collection<Listener> listeners;
+
+        private final String userAgent;
+        private final String elasticsearchVersion;
+
+        private boolean success = false;
+        private Exception exception = null;
+
+        private InputStream input = null;
+        private OutputStream output = null;
+        private URLConnection connection = null;
+
+        private static final int MAX_REDIRECTIONS = 5;
+        private Map<String, String> redirections = Maps.newHashMapWithExpectedSize(MAX_REDIRECTIONS);
+
+        DownloadThread(String source, Path destination, Collection<Listener> listeners, String userAgent, String elasticsearchVersion) {
+            this.source = source;
+            this.destination = destination;
+            this.listeners = listeners;
+            this.userAgent = userAgent;
+            this.elasticsearchVersion = elasticsearchVersion;
+        }
+
+        @Override
+        public void run() {
+            try {
+                success = download();
+            } catch (Exception e) {
+                exception = e;
+            }
+        }
+
+        private boolean download() throws IOException {
+            String current = source;
+
+            if (listeners != null) {
+                for (Listener listener : listeners) {
+                    listener.onDownloadBegin(current);
+                }
+            }
+
+            while ((current != null) && !isInterrupted()) {
+                try {
+                    URL url = URI.create(current).toURL();
+
+                    // Set up the URL connection
+                    connection = url.openConnection();
+                    if (connection == null) {
+                        return false;
+                    }
+
+                    connection.setUseCaches(true);
+                    connection.setConnectTimeout(5000);
```

Comment:
What will happen if we define a `TimeOut` in `DefaultPluginDownloader` which is superior at the one you set manually here. Should it be consistent?